File: src/sandbox.py
"""Per-job XDG sandbox dir helper.

Plan: parallel-readme-and-solo-filter Phase 2 (P7).

Each parallel claude job gets its own ``$SANDBOX/claude-jobs/<repo>/{config,data,cache,state}``
tree so concurrent runs don't race on the claude session DB.
"""
from __future__ import annotations

from pathlib import Path

from src import safety

_SUBDIRS = ("config", "data", "cache", "state")


def sandbox_for(base: Path, repo_name: str) -> dict[str, Path]:
    """Return ``{name: path}`` for the four XDG subdirs, creating them on disk.

    Repo name is validated to prevent path traversal.
    """
    safety.validate_repo_name(repo_name)
    root = base / "claude-jobs" / repo_name
    paths = {n: root / n for n in _SUBDIRS}
    for p in paths.values():
        p.mkdir(parents=True, exist_ok=True)
    return paths


def sandbox_env(paths: dict[str, Path]) -> dict[str, str]:
    """Return env-var dict mapping XDG_*_HOME → sandbox subdir path strings."""
    return {
        "XDG_CONFIG_HOME": str(paths["config"]),
        "XDG_DATA_HOME": str(paths["data"]),
        "XDG_CACHE_HOME": str(paths["cache"]),
        "XDG_STATE_HOME": str(paths["state"]),
    }

File: go/internal/sandbox/sandbox.go
// Package sandbox: per-job XDG sandbox dirs for parallel claude invocations.
package sandbox

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/salamientark/writeme/internal/safety"
)

// SubDirs are the four XDG sub-directories laid down per job.
var SubDirs = []string{"config", "data", "cache", "state"}

// Job represents a per-repo XDG sandbox tree.
type Job struct {
	Root  string
	Paths map[string]string
}

// JobSandbox creates <base>/claude-jobs/<repo>/{config,data,cache,state}.
func JobSandbox(base, repoName string) (*Job, error) {
	if err := safety.ValidateRepoName(repoName); err != nil {
		return nil, err
	}
	root := filepath.Join(base, "claude-jobs", repoName)
	paths := map[string]string{}
	for _, name := range SubDirs {
		p := filepath.Join(root, name)
		if err := os.MkdirAll(p, 0o755); err != nil {
			return nil, fmt.Errorf("mkdir %s: %w", p, err)
		}
		paths[name] = p
	}
	return &Job{Root: root, Paths: paths}, nil
}

// EnvFor returns the XDG_*_HOME overrides for the claude subprocess.
func EnvFor(job *Job) []string {
	return []string{
		"XDG_CONFIG_HOME=" + job.Paths["config"],
		"XDG_DATA_HOME=" + job.Paths["data"],
		"XDG_CACHE_HOME=" + job.Paths["cache"],
		"XDG_STATE_HOME=" + job.Paths["state"],
	}
}

// Cleanup removes the per-job sandbox tree (idempotent).
func (j *Job) Cleanup() error { return os.RemoveAll(j.Root) }
